File: python/basics/recursive_backtracking/212_word_search_II.py
'''
https://leetcode.com/problems/word-search-ii/

using trie.

1) create a trie of the words..
2) parameters: r, c, current-trie-node
    Choice: up, right, down, left

    end criteria: if the end has the word..

'''

def word_search_II(board: list[list[str]], words: list[str]) -> list[str]: 
    trie = {}
    rows = len(board)
    cols = len(board[0])
    WOD_KEY="$"
    result = []
    def add_words_to_trie():
        for word in words:
            node = trie
            for c in word:
                if c not in node:
                    node[c] = {}
                node = node[c]
                #node = node.setdefault(c, {})
            node[WOD_KEY] = word
    
    add_words_to_trie()
    #print("trie: " + str(trie))
    
    def is_valid(r,c):
        if r >=0 and r < rows and c >= 0 and c < cols:
            return True
        return False

    def backtrack(r, c, parent):
        letter = board[r][c]
      
        current_node = parent[letter]
        match = current_node.pop(WOD_KEY, False)
        if match:
            result.append(match[:])
        
        board[r][c] = "#"
        offsets = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        for off in offsets:
            row = r + off[0]
            col = c + off[1]
            if not is_valid(row, col):
                continue

            if not board[row][col] in current_node:
                continue
            backtrack( row, col, current_node)
        board[r][c] = letter
        
    #print("board-stat: " + str(board))
    for r in range(rows):
        for c in range(cols):
            #print("board[r][c]: " + str(board[r][c]))
            #print("trie: " + str(trie.keys()))
            
            if board[r][c] in trie:
                backtrack(r, c, trie)
    #print("board-end:  " + str(board))
    return result



board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]
words = ["oath","pea","eat","rain"]
words = ["pea","oath","eat","rain"]
#Output: ["eat","oath"]
result = word_search_II(board, words)
print("result: " + str(result))

exit(0)

board = [["a","b"],["c","d"]]
words = ["abcb"]
#Output: []
result = word_search_II(board, words)
print("result: " + str(result))

File: problem_solving/recursive_backtracking/cpp/212_word_search_II.cpp
/*
https://leetcode.com/problems/word-search-ii/
*/
#include<iostream>
#include<unordered_map>
#include<vector>
#include<string>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"

#include<iostream>
#include<vector>


using namespace std;

class TrieNode {
    public:
    std::string word;
    std::unordered_map<char, TrieNode*> children;

    TrieNode() {
    }
};  


class Trie {
    public:
    TrieNode* root;
    
    public:
    Trie() : root(new TrieNode()) {
    }

    void add_word(const std::string& word) {
        TrieNode* current = root;
        for(char ch : word) {
            if(current->children.find(ch) == current->children.end()) {
                current->children[ch] = new TrieNode();
            }
            current = current->children[ch];
        }
        if(current) {
            current->word = word;
        }
    }

    bool search(const std::string& word) {
        TrieNode* current = root;
        for(char ch : word) {
            if(current->children.find(ch) == current->children.end()) {
                return false;
            }
            current = current->children[ch];
        }
        return (current != nullptr && current->word == word);
    }
};


class Solution
{
private:
    int rows;
    int cols;

    void recur(std::vector<std::vector<char>>& board, int r, int c, TrieNode* current, std::vector<string>& results) {
        if( !is_valid(r, c)) {
            //std::cout << "return-1: " << r << "," << c << std::endl;
            return;
        }
        char ch = board[r][c];
        if(board[r][c]== '#') {
            //std::cout << "return-2: " << ch << std::endl;
            return;
        }
        
        //std::cout << "before-1: " << ch << std::endl;
        if( current->children.find(ch) == current->children.end() ) {
            //std::cout << "return-3: " << ch << std::endl;
            return;
        }
        //std::cout << "after-1" << std::endl;
        current = current->children[ch];
        if( !current->word.empty() ) {
            //std::cout << "word found" << std::endl;
            results.push_back(current->word);
            current->word = "";
        }       
        
        ch = board[r][c];
        board[r][c] = '#';

        for(const auto& tup : neighbors) {
            int row = r + std::get<0>(tup);
            int col = c + std::get<1>(tup);
            recur(board, row, col, current, results);
        }

        board[r][c] = ch;
    }

    std::vector<std::tuple<int, int>> neighbors = {{0, -1}, {-1, 0}, {0, 1}, {1, 0}};

    bool is_valid(int r, int c) {
        if(r < 0 || r >= rows || c < 0 || c >= cols ) {
            return false;
        }
        return true;
    }
public:

std::vector<string> findWords(std::vector<std::vector<char>>& board, const std::vector<std::string>& words)  {
    Trie trie;
    for(const auto& word : words) {
        trie.add_word(word);
    }
    //std::cout << trie.search("oath") << std::endl;
    rows = board.size();
    cols = 0;
    if(rows > 0) {
        cols = board[0].size();
    }
    std::vector<std::string> result;
    for(int r=0; r < rows; ++r) {
        for(int c=0; c<cols; ++c) {
            recur(board, r, c, trie.root, result);
        }
    }
    return result;
}
};

int main() {
    /*
    Trie trie;
    trie.add_word("apple");
    trie.add_word("mango");
    std::string ss = "apple";
    std::cout << trie.search(ss) << std::endl;
    std::cout << trie.search("app") << std::endl;
    std::cout << trie.search("apple1") << std::endl;
    std::cout << trie.search("mango") << std::endl;
    */
    std::vector<std::vector<char>> board = {
                                            {'o','a','a','n'},
                                            {'e','t','a','e'},
                                            {'i','h','k','r'},
                                            {'i','f','l','v'}};
    std::vector<std::string> words = {"oath","pea","eat","rain"};
    //Output: {'eat','oath'}
    Solution solv;
    auto result = solv.findWords( board, words);    
    print_vector(result);    
    

    board = {{'a','b'},{'c','d'}};
    words = {"abcb"};
    //Output: {}
    Solution solv2;
    result = solv2.findWords( board, words);    
    print_vector(result);
}